# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/h0pi/flyrank-ml-intenrship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

## 1. My lane and why

I am choosing the **Content Refresh Prioritization** lane.

The decision I want to improve is which existing content pages an SEO or content team should review first. I chose this lane because the starter dataset contains one row per pseudonymized content item and includes observable page-level signals such as impressions, CTR, average position, content age, word count, and days since the last update.

The intended output is not an automatic content-change system. It is a ranked review queue that helps a team focus limited editorial time on pages that may need attention. The starter notebooks showed that a fixed rule can rank pages, while a readable model can combine several signals. In later work, I will use an honest time-based outcome from the warehouse data rather than treating `trend_direction` or `trend_pct` as model features.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## 2. The question: decision, action, and cost of a wrong call

### Search question
Which existing content pages should an SEO or content team prioritize for review because they show signals consistent with declining search performance while still having meaningful visibility?

### Unit of analysis
One content page.

### Output
A ranked page-level review queue with a priority score. Higher-ranked pages should be reviewed earlier.

### Decision
A content or SEO manager decides which pages should be assigned for review first.

### Action
For each high-priority page, an editor or SEO specialist can inspect search intent, content freshness, title and meta description, internal linking, position, CTR, and potential technical issues. They can then decide whether to update, consolidate, redirect, or leave the page unchanged.

### Cost of a wrong recommendation
A false positive costs editor and SEO time on a page that was not an important opportunity. A false negative means a genuinely declining page may not be reviewed in time and could continue losing organic visibility or traffic.

### Why data or ML can help
A team cannot manually inspect every page consistently at scale. Multiple observable signals can be combined to rank pages more systematically than a single fixed rule. ML is useful only if it provides a more useful, honestly validated ranking than a transparent baseline rule.

### Task type and success metric
This is a ranking/scoring task because the decision is “which pages should be reviewed first?”. An initial metric is Precision@50: among the 50 pages ranked highest, what proportion truly have a later observed decline. The final target and validation will use time-aware warehouse data.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

## 3. Quick look at the data

The starter dataset contains 30,000 pseudonymized content items from 32 clients. I will use the following code to measure how many pages are currently marked as declining, how many still have meaningful search visibility, and how long pages have typically gone without an update.

These numbers do not prove that any feature causes decline. They show that the dataset contains enough page-level variation to make prioritization a useful decision-support problem.

In [6]:
import os
import sys
import subprocess
import pandas as pd

repo_url = "https://github.com/h0pi/flyrank-ml-intenrship.git"
repo_dir = "/content/flyrank-ml-intenrship"

if not os.path.exists(repo_dir):
    subprocess.run(
        ["git", "clone", "--depth", "1", repo_url, repo_dir],
        check=True
    )

os.chdir(repo_dir)

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

declining_rate = df["trend_direction"].str.lower().eq("down").mean()
visible_pages = (df["impressions_90d"] >= 500).sum()
median_days_since_update = df["days_since_last_update"].median()

print(f"Total pages: {len(df):,}")
print(f"Declining-page rate in the starter label: {declining_rate:.1%}")
print(f"Pages with at least 500 impressions in 90 days: {visible_pages:,}")
print(f"Median days since last update: {median_days_since_update:.0f}")

Total pages: 30,000
Declining-page rate in the starter label: 54.2%
Pages with at least 500 impressions in 90 days: 16,726
Median days since last update: 20


The starter dataset contains 30,000 pseudonymized content pages. In the starter label, 54.2% of pages are marked as declining, which shows that prioritization cannot be based only on a few obvious cases.

There are 16,726 pages with at least 500 impressions in the trailing 90 days. These pages still have meaningful search visibility, so choosing which ones to review first can affect where editorial effort is spent.

The median page was updated 20 days ago. This suggests that time since the last update alone is not enough to decide which pages should be reviewed; a useful ranking should combine multiple signals.

These are observed descriptive numbers from the starter dataset. They support the value of a page-level prioritization workflow, but they do not show that any individual feature causes search decline.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## 4. Careful words: what I can and cannot claim

This project can measure observed patterns associated with page performance and produce a ranked queue to support human review. It can describe which pages have signals that are associated with a later measured decline, using a time-aware validation approach when the warehouse data is available.

It cannot prove that content age, CTR, impressions, position, or any other single feature causes a search decline. It cannot reveal or predict Google's algorithm, and it cannot guarantee that updating a recommended page will improve traffic.

The final output is decision support for a content or SEO team. A human should inspect each recommendation and decide what action, if any, is appropriate.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.